src/dma/traits.rs

```diff
@@ -0,0 +1,1394 @@
+
+    /// Enable/disable peripheral increment (pinc) for the DMA stream.
+    fn set_peripheral_increment(&mut self, increment: bool);
+
+    /// Set the direction (dir) of the DMA stream.
+    fn set_direction<D: Direction>(&mut self, direction: D);
+
+    /// Convenience method to configure the 4 common interrupts for the DMA stream.
+    fn set_interrupts_enable(
+        &mut self,
+        transfer_complete: bool,
+        half_transfer: bool,
+        transfer_error: bool,
+        direct_mode_error: bool,
+    );
+
+    /// Enable/disable the transfer complete interrupt (tcie) of the DMA stream.
+    fn set_transfer_complete_interrupt_enable(&mut self, transfer_complete_interrupt: bool);
+
+    /// Enable/disable the half transfer interrupt (htie) of the DMA stream.
+    fn set_half_transfer_interrupt_enable(&mut self, half_transfer_interrupt: bool);
+
+    /// Enable/disable the transfer error interrupt (teie) of the DMA stream.
+    fn set_transfer_error_interrupt_enable(&mut self, transfer_error_interrupt: bool);
+
+    /// Enable/disable the direct mode error interrupt (dmeie) of the DMA stream.
+    fn set_direct_mode_error_interrupt_enable(&mut self, direct_mode_error_interrupt: bool);
+
+    /// Enable/disable the fifo error interrupt (feie) of the DMA stream.
+    fn set_fifo_error_interrupt_enable(&mut self, fifo_error_interrupt: bool);
+
+    /// Enable/disable the double buffer (dbm) of the DMA stream.
+    fn set_double_buffer(&mut self, double_buffer: bool);
+
+    /// Set the fifo threshold (fcr.fth) of the DMA stream.
+    fn set_fifo_threshold(&mut self, fifo_threshold: config::FifoThreshold);
+
+    /// Enable/disable the fifo (dmdis) of the DMA stream.
+    fn set_fifo_enable(&mut self, fifo_enable: bool);
+
+    /// Set memory burst mode (mburst) of the DMA stream.
+    fn set_memory_burst(&mut self, memory_burst: config::BurstMode);
+
+    /// Set peripheral burst mode (pburst) of the DMA stream.
+    fn set_peripheral_burst(&mut self, peripheral_burst: config::BurstMode);
+
+    /// Get the current fifo level (fs) of the DMA stream.
+    fn fifo_level() -> FifoLevel;
+
+    /// Get which buffer is currently in use by the DMA.
+    fn current_buffer() -> CurrentBuffer;
+}
+
+/// DMA direction.
+pub trait Direction: Bits<u8> {
+    /// Creates a new instance of the type.
+    fn new() -> Self;
+
+    /// Returns the `DmaDirection` of the type.
+    fn direction() -> DmaDirection;
+}
+
+/// Get an address and memory size the DMA can use.
+pub trait PeriAddress: Sealed {
+    /// Memory size of the peripheral.
+    type MemSize;
+
+    /// Returns the address to be used by the DMA stream.
+    fn address(&self) -> u32;
+}
+
+// Convenience macro for implementing addresses on peripherals
+macro_rules! address {
+    ($(($peripheral:ty, $register:ident, $size: ty)),+ $(,)*) => {
+        $(
+            impl Sealed for $peripheral {}
+
+            impl PeriAddress for $peripheral {
+                #[inline]
```

Comment:
Hmm, not sure about the impact, first I thought of `#[inline(always)]` which I think would cause inlining in debug builds, but I thought maybe it would be too harsh ? not sure.

Reply:
I usually only add compiler hints only after actually testing if they improve anything (inlining can have negative effects on caching). But if there is an a convention how this is done in this crate or the embedded ecosystem in general, it's probably best to go with that. I don't mean to start a huge discussion here :)